plugin/src/main/java/org/opensearch/ml/autoreload/MLModelAutoReDeployer.java

```diff
@@ -0,0 +1,269 @@
+                }
+            }
+        }, e -> { log.error("Failed to query need undeploy models, no action will be performed"); });
+        queryRunningModels(listener);
+    }
+
+    private void queryRunningModels(ActionListener<SearchResponse> listener) {
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        QueryBuilder queryBuilder = QueryBuilders
+            .boolQuery()
+            .mustNot(QueryBuilders.existsQuery(MLModel.CHUNK_NUMBER_FIELD))
+            .should(QueryBuilders.termQuery(MLModel.MODEL_STATE_FIELD, MLModelState.DEPLOYING.name()))
+            .should(QueryBuilders.termQuery(MLModel.MODEL_STATE_FIELD, MLModelState.DEPLOYED.name()))
+            .should(QueryBuilders.termQuery(MLModel.MODEL_STATE_FIELD, MLModelState.PARTIALLY_DEPLOYED.name()));
+
+        FieldSortBuilder sortBuilder = SortBuilders.fieldSort(MLModel.LAST_DEPLOYED_TIME_FIELD).order(SortOrder.ASC);
+
+        String[] includes = new String[] { MLModel.AUTO_REDEPLOY_RETRY_TIMES_FIELD, MLModel.PLANNING_WORKER_NODES_FIELD };
+
+        String[] excludes = new String[] { MLModel.MODEL_CONTENT_FIELD, MLModel.OLD_MODEL_CONTENT_FIELD };
+        FetchSourceContext fetchContext = new FetchSourceContext(true, includes, excludes);
+        searchSourceBuilder.query(queryBuilder).sort(sortBuilder).fetchSource(fetchContext);
+        SearchRequestBuilder searchRequestBuilder = new SearchRequestBuilder(client, SearchAction.INSTANCE)
+            .setIndices(ML_MODEL_INDEX)
+            .setSource(searchSourceBuilder);
+
+        searchRequestBuilder.execute(listener);
+    }
+
+    @SuppressWarnings("unchecked")
+    private void triggerModelRedeploy(ModelAutoRedeployArrangement modelAutoRedeployArrangement) {
+        String modelId = modelAutoRedeployArrangement.getSearchResponse().getId();
+        List<String> addedNodes = modelAutoRedeployArrangement.getAddedNodes();
+        List<String> planningWorkerNodes = (List<String>) modelAutoRedeployArrangement
+            .getSearchResponse()
+            .getSourceAsMap()
+            .get(MLModel.PLANNING_WORKER_NODES_FIELD);
+        Integer autoRedeployRetryTimes = (Integer) modelAutoRedeployArrangement
+            .getSearchResponse()
+            .getSourceAsMap()
+            .get(MLModel.AUTO_REDEPLOY_RETRY_TIMES_FIELD);
+        // calculate node ids.
+        String[] nodeIds = null;
+        if (allowCustomDeploymentPlan && planningWorkerNodes != null && planningWorkerNodes.size() > 0) {
+            List<String> needRedeployPlanningWorkerNodes = Arrays
+                .stream(planningWorkerNodes.toArray(new String[0]))
+                .filter(addedNodes::contains)
+                .collect(Collectors.toList());
+            nodeIds = needRedeployPlanningWorkerNodes.size() > 0 ? planningWorkerNodes.toArray(new String[0]) : null;
+        } else {
+            nodeIds = new String[0];
+        }
+
+        if (nodeIds == null) {
```

Comment:
Line 227 will set nodeIds as `new String[0];` , so will never enter this if branch ?

Reply:
No, either allowCustomDeploymentPlan is false or (allowCustomDeploymentPlan is true but planningWorkerNodes is empty), will enter else branch, in both cases, we send the request without node ids.